street2sat_utils/client.py

```diff
@@ -216,28 +204,59 @@ def get_new_points(time_dict: Dict, coord_dict: Dict, distance_dict: Dict) -> Tu
         initial_bearing = math.degrees(initial_bearing)
         compass_bearing = (initial_bearing + 360) % 360
 
-        if closest_time < time_val:
+        if closest.time < self.time:
             compass_bearing = (compass_bearing + 180) % 360
 
         ##### IMPORTANT ASSUMPTION ALL CAMERAS POINT LEFT ####
-        bearings[file] = (compass_bearing - 90) % 360
-
-    # file:crop:(lat,long)
-    new_points = {}
-    for file, bearing in bearings.items():
-        if file in distance_dict.keys():
-            distance_meters = distance_dict[file]
-            translated_coords = point_meters_away(
-                coord_dict[file], bearing, distance_meters
+        return (compass_bearing - 90) % 360
+
+    def set_crop_coord(self, closest: Prediction) -> None:
+        distances = self.generate_distances()
+        bearing = self.compute_bearing(closest)
+
+        lat1 = math.radians(self.coord[0])  # Current lat point converted to radians
+        lon1 = math.radians(self.coord[1])  # Current long point converted to radians
+
+        self.crop_coord = {}
+
+        for crop, meters in distances.items():
+            R = 6378.1  # Radius of the Earth
+            brng = math.radians(bearing)  # Bearing is degrees converted to radians.
+            d = meters / 1000  # Distance in km
+
+            lat2 = math.asin(
+                math.sin(lat1) * math.cos(d / R)
+                + math.cos(lat1) * math.sin(d / R) * math.cos(brng)
+            )
+            lon2 = lon1 + math.atan2(
+                math.sin(brng) * math.sin(d / R) * math.cos(lat1),
+                math.cos(d / R) - math.sin(lat1) * math.sin(lat2),
             )
-            new_points[file] = translated_coords
 
-    return bearings, new_points
+            lat2 = math.degrees(lat2)
+            lon2 = math.degrees(lon2)
+
+            self.crop_coord[crop] = (lat2, lon2)
 
 
-if __name__ == "__main__":
-    model = get_model("../street2sat_utils/model_weights/best.pt")
-    img1_path = "../example_images/GP__1312.JPG"
-    img1 = cv2.cvtColor(cv2.imread(img1_path), cv2.COLOR_BGR2RGB)
-    run_prediction(img1, model)
-    # predict('../temp/uQBYaAsIgV/')
+def calculate_crop_coords(preds: List[Prediction]) -> List[Prediction]:
```

Comment:
This function replaces the previous `get_new_points` and is more efficient by first sorting the predictions by time.
There was actually a bug in the previous implementation where predictions after were always the closest